src/looprestoration.rs

```diff
@@ -934,3 +955,194 @@ unsafe fn sgr_mix_rust<BD: BitDepth>(
         j += 1;
     }
 }
+
+// TODO(randomPoison): Temporarily pub until all usages can be made private.
+#[cfg(all(feature = "asm", any(target_arch = "arm", target_arch = "aarch64")))]
+pub(crate) trait BitDepthLooprestorationArm: BitDepth {
+    fn dav1d_sgr_box3_h_neon(
+        sumsq: *mut int32_t,
+        sum: *mut int16_t,
+        left: *const [Self::Pixel; 4],
+        src: *const Self::Pixel,
+        stride: ptrdiff_t,
+        w: libc::c_int,
+        h: libc::c_int,
+        edges: LrEdgeFlags,
+    );
+
+    fn dav1d_sgr_finish_filter1_neon(
+        tmp: *mut int16_t,
+        src: *const Self::Pixel,
+        stride: ptrdiff_t,
+        a: *const int32_t,
+        b: *const int16_t,
+        w: libc::c_int,
+        h: libc::c_int,
+    );
+}
+
+#[cfg(all(feature = "asm", any(target_arch = "arm", target_arch = "aarch64")))]
+impl BitDepthLooprestorationArm for BitDepth8 {
+    #[inline(always)]
+    fn dav1d_sgr_box3_h_neon(
+        sumsq: *mut int32_t,
+        sum: *mut int16_t,
+        left: *const [Self::Pixel; 4],
+        src: *const Self::Pixel,
+        stride: ptrdiff_t,
+        w: libc::c_int,
+        h: libc::c_int,
+        edges: LrEdgeFlags,
+    ) {
+        extern "C" {
+            fn dav1d_sgr_box3_h_8bpc_neon(
+                sumsq: *mut int32_t,
+                sum: *mut int16_t,
+                left: *const [<BitDepth8 as BitDepth>::Pixel; 4],
+                src: *const <BitDepth8 as BitDepth>::Pixel,
```

Comment:
I thought we were gonna make these type erased so that we didn't need to do the bitdepth trait thing.

Reply:
In this case since the `extern "C"` declaration is inside of a bitdepth-specific trait impl, we don't need to do type erasure. I wish I could do that as `Self::Pixel` instead of `<BitDepth8 as BitDepth>::Pixel`, but Rust doesn't allow the former (because reasons). Not doing type erasure here doesn't matter much, though, it just saves a couple of extra casts 🤷

Reply:
Wouldn't type erasure fix the need to have a whole trait and impls?

Let me look more closely at the PR, though.  I've only looked at it cursorily so far.

Reply:
The main reason for the trait is that we need to call functions with bitdepth-dependent names from a generic context. Type erasure is more related to being able to store function pointers in a bitdepth-generic way.